File: imu_calib_matlab/src/tasks/run_acc_calibration.m
function task = run_acc_calibration(inputData, varargin)
%RUN_ACC_CALIBRATION Run accelerometer calibration using the new workflow.
% Accepted inputs:
%   - accPoses / acc_poses: pre-computed pose means
%   - static: continuous raw data; static segments will be extracted automatically

opts = parse_options(varargin{:});
[accPoses, extractionInfo, source] = normalize_acc_input(inputData, opts.options);

if isempty(accPoses)
    task = make_task_result(false, ...
        'Accelerometer calibration requires accPoses / acc_poses or raw static data with detectable static segments.', [], ...
        'missing_inputs', {'accPoses_or_static_segments'}, ...
        'meta', struct('task_name', 'run_acc_calibration', 'source', source));
    return;
end

if ~isempty(extractionInfo)
    minSegments = opts.options.segmentation.min_static_segments;
    numSegments = extractionInfo.num_segments;
    if numSegments < minSegments
        task = make_task_result(false, ...
            sprintf(['Not enough static segments were extracted for accelerometer calibration. ' ...
                     'Detected %d, but at least %d are recommended.'], numSegments, minSegments), [], ...
            'warnings', extractionInfo.messages, ...
            'missing_inputs', {'static_segments'}, ...
            'meta', struct('task_name', 'run_acc_calibration', 'source', source));
        return;
    end
end

try
    [Ca, ba, fitInfo] = fit_acc_multi_pose(accPoses, ...
        'gravity_magnitude', opts.options.acc_calibration.gravity_magnitude, ...
        'options', opts.options);
catch ME
    warnList = {};
    if ~isempty(extractionInfo)
        warnList = extractionInfo.messages;
    end
    task = make_task_result(false, ME.message, [], ...
        'warnings', warnList, ...
        'meta', struct('task_name', 'run_acc_calibration', 'source', source));
    return;
end

if ~isempty(extractionInfo)
    fitInfo.static_segment_extraction = extractionInfo;
end

warnings = {};
if ~isempty(extractionInfo)
    warnings = [warnings, extractionInfo.messages]; %#ok<AGROW>
end
if isfield(fitInfo, 'warnings') && ~isempty(fitInfo.warnings)
    warnings = [warnings, fitInfo.warnings(:).']; %#ok<AGROW>
end
if detect_legacy_refs(accPoses)
    warnings{end + 1} = ...
        'Legacy reference vectors were detected. They are deprecated and were only used for initialization compatibility.'; %#ok<AGROW>
end

result = struct();
result.Ca = Ca;
result.ba = ba;
result.Sa = fitInfo.Sa;
result.Ma = fitInfo.Ma;
result.gravity_magnitude = fitInfo.gravity_magnitude;
result.fitInfo = fitInfo;

task = make_task_result(true, 'accelerometer calibration completed successfully.', result, ...
    'warnings', warnings, ...
    'meta', struct('task_name', 'run_acc_calibration', 'source', source));
end

function opts = parse_options(varargin)
opts = struct();
opts.options = default_calib_options();
if mod(numel(varargin), 2) ~= 0
    error('run_acc_calibration:InvalidInput', ...
        'Optional arguments must be provided as name/value pairs.');
end
for k = 1:2:numel(varargin)
    name = lower(char(varargin{k}));
    value = varargin{k + 1};
    switch name
        case 'options'
            opts.options = value;
        otherwise
            error('run_acc_calibration:UnknownOption', ...
                'Unknown option "%s".', char(varargin{k}));
    end
end
end

function [accPoses, extractionInfo, source] = normalize_acc_input(inputData, options)
accPoses = [];
extractionInfo = [];
source = 'direct_input';

if isstruct(inputData)
    if isfield(inputData, 'accPoses') && ~isempty(inputData.accPoses)
        accPoses = inputData.accPoses;
        source = 'dataset.accPoses';
        return;
    end
    if isfield(inputData, 'acc_poses') && ~isempty(inputData.acc_poses)
        accPoses = inputData.acc_poses;
        source = 'dataset.acc_poses';
        return;
    end
    if isfield(inputData, 'static') && ~isempty(inputData.static) && isstruct(inputData.static)
        [accPoses, extractionInfo] = extract_static_pose_means(inputData.static, 'options', options);
        source = 'dataset.static';
        return;
    end
end

if numel(inputData) > 0 && isstruct(inputData)
    accPoses = inputData;
end
end

function tf = detect_legacy_refs(accPoses)
tf = false;
for i = 1:numel(accPoses)
    if isfield(accPoses(i), 'a_ref') && ~isempty(accPoses(i).a_ref)
        tf = true;
        return;
    end
end
end

File: imu_calib_python/imu_calib/tasks/run_acc_calibration.py
from __future__ import annotations

from dataclasses import is_dataclass

from imu_calib.calib.extract_static_pose_means import extract_static_pose_means
from imu_calib.calib.fit_acc_multi_pose import fit_acc_multi_pose
from imu_calib.runtime.default_calib_options import default_calib_options
from imu_calib.tasks.make_task_result import make_task_result
from imu_calib.utils.exceptions import ImuCalibError


def run_acc_calibration(input_data, *, options: dict | None = None) -> dict:
    """Run accelerometer calibration using the new gravity-constraint workflow.

    Accepted inputs:
    - acc_poses: pre-computed per-pose static means (new preferred path)
    - static: continuous raw IMU data; static segments will be detected and
      converted into pose means automatically

    Legacy reference vectors a_ref are optional and only used for initialization
    compatibility. The optimization objective no longer depends on them.
    """

    options = default_calib_options() if options is None else options
    acc_poses, extraction_info, source = _normalize_acc_input(input_data, options)
    if not acc_poses:
        return make_task_result(
            False,
            "Accelerometer calibration requires acc_poses or raw static data with detectable static segments.",
            None,
            missing_inputs=["acc_poses_or_static_segments"],
            meta={"task_name": "run_acc_calibration", "source": source},
        )

    if extraction_info is not None:
        min_segments = int(options["segmentation"]["min_static_segments"])
        num_segments = int(extraction_info.get("num_segments", 0))
        if num_segments < min_segments:
            return make_task_result(
                False,
                (
                    "Not enough static segments were extracted for accelerometer calibration. "
                    f"Detected {num_segments}, but at least {min_segments} are recommended."
                ),
                None,
                warnings=extraction_info.get("messages", []),
                missing_inputs=["static_segments"],
                meta={"task_name": "run_acc_calibration", "source": source},
            )

    try:
        Ca, ba, fit_info = fit_acc_multi_pose(
            acc_poses,
            gravity_magnitude=options["acc_calibration"]["gravity_magnitude"],
            options=options,
        )
    except ImuCalibError as exc:
        return make_task_result(
            False,
            str(exc),
            None,
            warnings=extraction_info["messages"] if extraction_info is not None else [],
            meta={"task_name": "run_acc_calibration", "source": source},
        )
    if extraction_info is not None:
        fit_info["static_segment_extraction"] = extraction_info

    warnings: list[str] = []
    if extraction_info is not None:
        warnings.extend(extraction_info.get("messages", []))
    warnings.extend(list(fit_info.get("warnings", [])))
    if any(getattr(pose, "a_ref", None) is not None for pose in acc_poses):
        warnings.append(
            "Legacy reference vectors were detected. They are deprecated and were only used for initialization compatibility."
        )

    return make_task_result(
        True,
        "accelerometer calibration completed successfully.",
        {
            "Ca": Ca,
            "ba": ba,
            "Sa": fit_info["Sa"],
            "Ma": fit_info["Ma"],
            "gravity_magnitude": fit_info["gravity_magnitude"],
            "fitInfo": fit_info,
        },
        warnings=warnings,
        meta={"task_name": "run_acc_calibration", "source": source},
    )


def _normalize_acc_input(input_data, options: dict):
    acc_poses = []
    extraction_info = None
    source = "direct_input"

    if is_dataclass(input_data):
        if hasattr(input_data, "acc_poses") and input_data.acc_poses:
            acc_poses = input_data.acc_poses
            source = "dataset.acc_poses"
        elif hasattr(input_data, "static") and input_data.static is not None:
            acc_poses, extraction_info = extract_static_pose_means(input_data.static, options=options)
            source = "dataset.static"
    elif isinstance(input_data, dict):
        if input_data.get("acc_poses"):
            acc_poses = input_data["acc_poses"]
            source = "dict.acc_poses"
        elif input_data.get("static") is not None:
            acc_poses, extraction_info = extract_static_pose_means(input_data["static"], options=options)
            source = "dict.static"
    else:
        acc_poses = input_data

    return acc_poses, extraction_info, source
